### PR TITLE
Outlier detect

### PR DESCRIPTION
**Pull request by tangyl92**

### CODE REVIEW

**Comment by tangyl92** [comment]
create function, Since I don't know the sample data distribution and not sure if I can detect the outlier. I change the function to find the best and worst market campaign in given data based on chosen metric.

File: src/dscigametrics/metric4.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+def find_best_and_worst_campaigns(df, start_date, end_date, campaign_ids, metric):
+    """
+    Analyzes marketing campaign performance over a specified date range and identifies the best and worst performing campaigns based on a selected metric.
```
Comment (ianm99) [inline src/dscigametrics/metric4.py]: initial blank line after """

File: src/dscigametrics/metric4.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    This function iterates over a list of campaign IDs, calculates performance metrics for each campaign within the given date range, and then determines the best and worst performing campaigns based on the specified metric.
+
+    Parameters:
```
Comment (ianm99) [inline src/dscigametrics/metric4.py]: should insert a line like --------- under Parameters. Look at my docstring in compute_metrics.py

File: src/dscigametrics/metric4.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    df (pandas.DataFrame): A DataFrame containing campaign data. Expected to include columns for dates, campaign IDs, and various metrics like visits, transactions, and revenue.
+    start_date (datetime): The start date for the analysis period.
+    end_date (datetime): The end date for the analysis period.
+    campaign_ids (list of str): A list of campaign IDs to be analyzed.
+    metric (str): The name of the metric to be used for evaluating campaign performance. This should be one of the keys returned by the 'calculate_campaign_metrics' function, such as 'conversion_rate', 'total_transaction_revenue', etc.
+
+    Returns:
```
Comment (ianm99) [inline src/dscigametrics/metric4.py]: insert another ------- under Returns

File: src/dscigametrics/metric4.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    dict: A dictionary containing two key-value pairs - 'best_campaign' and 'worst_campaign'. Each is a dictionary itself, containing the 'campaign_id' and its corresponding 'metrics', which include the performance metrics calculated for that campaign.
+
+    Example usage:
```
Comment (ianm99) [inline src/dscigametrics/metric4.py]: examples not needed for now

File: src/dscigametrics/metric4.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    campaign_ids = ['campaign1', 'campaign2', 'campaign3'];
+    result = find_best_and_worst_campaigns(df, pd.to_datetime('2022-01-01'), pd.to_datetime('2022-03-31'), campaign_ids, 'conversion_rate');
+    print(result)
+    """
```
Comment (ianm99) [inline src/dscigametrics/metric4.py]: blank line before the """

File: src/dscigametrics/metric4.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+def find_best_and_worst_campaigns(df, start_date, end_date, campaign_ids, metric):
+    """
+    Analyzes marketing campaign performance over a specified date range and identifies the best and worst performing campaigns based on a selected metric.
+    This function iterates over a list of campaign IDs, calculates performance metrics for each campaign within the given date range, and then determines the best and worst performing campaigns based on the specified metric.
+
+    Parameters:
+    df (pandas.DataFrame): A DataFrame containing campaign data. Expected to include columns for dates, campaign IDs, and various metrics like visits, transactions, and revenue.
```
Comment (ianm99) [inline src/dscigametrics/metric4.py]: delete second sentence for now

File: src/dscigametrics/metric4.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    start_date (datetime): The start date for the analysis period.
+    end_date (datetime): The end date for the analysis period.
+    campaign_ids (list of str): A list of campaign IDs to be analyzed.
+    metric (str): The name of the metric to be used for evaluating campaign performance. This should be one of the keys returned by the 'calculate_campaign_metrics' function, such as 'conversion_rate', 'total_transaction_revenue', etc.
```
Comment (ianm99) [inline src/dscigametrics/metric4.py]: delete second sentence for now. Your information presupposes facts not in evidence

**Review by ianm99 — CHANGES_REQUESTED** [review]
Docstring needs a little work.

**Comment by tangyl92** [comment]
done

**Review by ianm99 — APPROVED** [review]
thanks!

File: src/dscigametrics/metric4.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        A list of campaign IDs to be analyzed.
+    metric : str
+        The name of the metric to be used for evaluating campaign performance. This should be one of the keys returned by the 'calculate_campaign_metrics' function, such as 'conversion_rate', 'total_transaction_revenue', etc.
+
+    Returns
+    ----------
+    dict
+        A dictionary containing two key-value pairs - 'best_campaign' and 'worst_campaign'. Each is a dictionary itself, containing the 'campaign_id' and its corresponding 'metrics', which include the performance metrics calculated for that campaign.
+
+
```
Comment (WeilinHan8) [inline src/dscigametrics/metric4.py:24]: Only one blank line before ```

File: src/dscigametrics/metric4.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+        The end date for the analysis period.
+    campaign_ids : list of str
+        A list of campaign IDs to be analyzed.
+    metric : str
+        The name of the metric to be used for evaluating campaign performance. This should be one of the keys returned by the 'calculate_campaign_metrics' function, such as 'conversion_rate', 'total_transaction_revenue', etc.
+
+    Returns
+    ----------
+    dict
+        A dictionary containing two key-value pairs - 'best_campaign' and 'worst_campaign'. Each is a dictionary itself, containing the 'campaign_id' and its corresponding 'metrics', which include the performance metrics calculated for that campaign.
```
Comment (WeilinHan8) [inline src/dscigametrics/metric4.py:22]: Please break the line manually, currently they are in one line and reader has to scroll to the far right to finish the sentence.

File: src/dscigametrics/metric4.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+def find_best_and_worst_campaigns(df, start_date, end_date, campaign_ids, metric):
+    '''
+
+    Analyzes and identifies the best and worst performing marketing campaigns based on a selected metric from Google Analytics data over a specified date range.
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+        A DataFrame containing campaign data. Expected to include columns for dates, campaign IDs, and various metrics like visits, transactions, and revenue.
```
Comment (WeilinHan8) [inline src/dscigametrics/metric4.py:9]: manually break this sentence to new line. Thanks

File: src/dscigametrics/metric4.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    start_date : datetime
+        The start date for the analysis period.
+    end_date : datetime
+        The end date for the analysis period.
+    campaign_ids : list of str
+        A list of campaign IDs to be analyzed.
+    metric : str
+        The name of the metric to be used for evaluating campaign performance. This should be one of the keys returned by the 'calculate_campaign_metrics' function, such as 'conversion_rate', 'total_transaction_revenue', etc.
```
Comment (WeilinHan8) [inline src/dscigametrics/metric4.py:17]: break the line manually so reader don't have to scroll to the right.

File: src/dscigametrics/metric4.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
+def find_best_and_worst_campaigns(df, start_date, end_date, campaign_ids, metric):
```
Comment (WeilinHan8) [inline src/dscigametrics/metric4.py:1]: please rename your file to your function name "find_best...." and make sure the old file was deleted.

**Review by WeilinHan8 — CHANGES_REQUESTED** [review]
Please make the changes I listed in file

File: src/dscigametrics/metric4.py [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    '''
+
+    Analyzes and identifies the best and worst performing marketing campaigns based on a selected metric from Google Analytics data over a specified date range.
+
```
Comment (WeilinHan8) [inline src/dscigametrics/metric4.py:5]: Also add the four metrics description in here, you can copy from Ian's script (compute_metrics)

**Comment by tangyl92** [comment]
@WeilinHan8 I reject them all.

**Pull request closed without merging**